File: 프로그래머스/5/49190. 방의 개수/방의 개수.py
from collections import defaultdict
# 방이 되는 기준 -> 이미 지나왔던 경로를 다시 지나갈 때 
# 이때 a-> b 이후 다시 b -> a로 가는 경우는 a가 다시 겹치지만 방이 생성되지 않음
# 그러므로 시작점과 도착점을 함께 봐야 함
# 그리고 대각선 교차를 측정하기 위해서는 좌표값을 2배로 해야 가능

def solution(arrows):
    # 일반 좌표계 기준으로 방향 설정 0 ~ 7
    direct = [[0, 1], [1, 1], [1, 0], [1, -1], [0, -1], [-1, -1], [-1, 0], [-1, 1]]
    visited = defaultdict(set)
    x, y = 0, 0
    room = 0
    
    for d in arrows:
        for _ in range(2): # 2배 좌표계 확장 대각선 교차도 확인해야 해서
            dx, dy = direct[d]
            nx, ny = x + dx, y + dy
            
            if (nx, ny) in visited and (x, y) not in visited[(nx, ny)]:
                room += 1 # 이전 nx, ny 방문한 적은 있지만 그게 x, y에서 온 것이 아니면 방 생성
            
            visited[(x, y)].add((nx, ny))
            visited[(nx, ny)].add((x, y))
            x, y = nx, ny
    
    return room

File: 프로그래머스/5/49190. 방의 개수/방의 개수.java
import java.util.*;

class Solution {
    
    static class Point{
        int x, y;
        
        public Point(int x, int y){
            this.x = x;
            this.y = y;
        }
        
        @Override 
        public boolean equals(Object obj){
            if(!(obj instanceof Point)) return false;
            Point p = (Point) obj;
            return this.x == p.x && this.y == p.y;
        }
        
        @Override
        public int hashCode(){
            return Objects.hash(x, y);
        }
    }
    
    
    public int solution(int[] arrows) {
        int[] dx = {0, 1, 1, 1, 0, -1, -1, -1};
        int[] dy = {1, 1, 0, -1, -1, -1, 0, 1};
        
        Set<Point> visited = new HashSet<>();
        Map<Point, Set<Point>> graph = new HashMap<>();
        
        Point current = new Point(0, 0);
        visited.add(current);
        
        int rooms = 0;
        for(int dir : arrows){
            for(int step = 0; step < 2; step++){ // 선 겹침 위해서 2회 진행
                Point next = new Point(current.x + dx[dir], current.y + dy[dir]);
                
                graph.putIfAbsent(current, new HashSet<>()); // 없으면 생성
                graph.putIfAbsent(next, new HashSet<>());
                
                if(visited.contains(next) && !graph.get(current).contains(next)){
                    rooms++;
                }
                
                graph.get(current).add(next);
                graph.get(next).add(current);
                
                visited.add(next);
                current = next;
            }
        }
        
        return rooms;
    }
}